File: scraper/flashscore_parser.py
import datetime
import re
from typing import Dict, List, Any, Optional

def parse_flashscore_feed(raw_text: str) -> List[Dict[str, Any]]:
    """
    Parse Flashscore's raw delimited feed (f_1_... format).
    Blocks are separated by '¬~', key-value pairs by '¬', and keys/values by '÷'.
    """
    blocks = raw_text.split('¬~')
    matches = []
    
    current_league = {
        'id': 'unknown',
        'name': 'Football Match',
        'country': 'World',
        'country_code': '',
    }

    for block in blocks:
        if not block.strip():
            continue
        
        parts = block.split('¬')
        data = {}
        for p in parts:
            if '÷' in p:
                k, v = p.split('÷', 1)
                data[k] = v

        # Check if tournament header block
        if 'ZA' in data:
            raw_za = data.get('ZA', '')
            country = data.get('ZY', 'World')
            league_name = raw_za
            
            # Often format is "COUNTRY: League Name"
            if ':' in raw_za:
                parts_za = raw_za.split(':', 1)
                country = parts_za[0].strip()
                league_name = parts_za[1].strip()

            current_league = {
                'id': data.get('ZC', data.get('ZEE', 'league_' + re.sub(r'[^a-zA-Z0-9]', '_', league_name.lower()))),
                'name': league_name,
                'country': country,
                'country_code': data.get('ZB', ''),
                'flag': data.get('OAJ', ''),
            }

        # Check if match block
        elif 'AA' in data:
            match_id = data.get('AA')
            home_name = data.get('AE', 'Home Team')
            away_name = data.get('AF', 'Away Team')
            
            # Scores
            raw_home_score = data.get('AG', '-')
            raw_away_score = data.get('AH', '-')
            home_score = int(raw_home_score) if raw_home_score.isdigit() else 0
            away_score = int(raw_away_score) if raw_away_score.isdigit() else 0

            # Match status mapping
            # AB: 1 = Scheduled, 2 = Live / In progress, 3 = Finished, 4 = Postponed, 5 = Cancelled, etc.
            status_code = data.get('AB', '1')
            status_text = data.get('AC', '')
            status = 'SCHEDULED'
            minute: Optional[int] = None

            if status_code == '2':
                # In play or paused
                if 'HT' in status_text or 'Half' in status_text:
                    status = 'PAUSED'
                elif 'ET' in status_text:
                    status = 'EXTRA_TIME'
                elif 'Pen' in status_text:
                    status = 'PENALTIES'
                else:
                    status = 'IN_PLAY'
                
                # Extract minute if available
                min_match = re.search(r'(\d+)', status_text)
                if min_match:
                    minute = int(min_match.group(1))
            elif status_code == '3':
                status = 'FINISHED'
                if not status_text:
                    status_text = 'Finished'
            elif status_code in ('4', '9'):
                status = 'POSTPONED'
            elif status_code in ('5', '6'):
                status = 'CANCELLED'
            elif status_code == '1':
                status = 'SCHEDULED'
                if not status_text and 'AD' in data:
                    try:
                        ts = int(data['AD'])
                        status_text = datetime.datetime.fromtimestamp(ts, tz=datetime.timezone.utc).strftime('%H:%M')
                    except:
                        status_text = 'Upcoming'

            # Start time
            start_iso = datetime.datetime.now(datetime.timezone.utc).isoformat()
            if 'AD' in data:
                try:
                    start_ts = int(data['AD'])
                    start_iso = datetime.datetime.fromtimestamp(start_ts, tz=datetime.timezone.utc).isoformat()
                except:
                    pass

            matches.append({
                'id': match_id,
                'provider': 'flashscore',
                'league': {
                    'id': current_league['id'],
                    'name': current_league['name'],
                    'country': current_league['country'],
                    'countryCode': current_league['country_code'],
                    'flag': current_league.get('flag', ''),
                },
                'homeTeam': {
                    'id': data.get('CX', f"team_{re.sub(r'[^a-zA-Z0-9]', '_', home_name.lower())}"),
                    'name': home_name,
                },
                'awayTeam': {
                    'id': data.get('CY', f"team_{re.sub(r'[^a-zA-Z0-9]', '_', away_name.lower())}"),
                    'name': away_name,
                },
                'homeScore': home_score,
                'awayScore': away_score,
                'status': status,
                'statusText': status_text or ('Live' if status == 'IN_PLAY' else status),
                'minute': minute,
                'startTime': start_iso,
                'lastUpdated': datetime.datetime.now(datetime.timezone.utc).isoformat(),
            })

    return matches

def parse_flashscore_events(match_id: str, raw_text: str) -> List[Dict[str, Any]]:
    """
    Parse Flashscore's match summary/incidents feed (df_su_1_... format).
    """
    blocks = raw_text.split('¬~')
    events = []

    for block in blocks:
        if not block.strip():
            continue
        parts = block.split('¬')
        data = {}
        for p in parts:
            if '÷' in p:
                k, v = p.split('÷', 1)
                data[k] = v

        # Key incident markers:
        # IB = Minute (e.g. 45+3' or 62')
        # IA = Team side (1 = Home, 2 = Away)
        # IF = Player name
        # IK = Incident description / type (e.g. "Goal", "Penalty", "Yellow Card", "Substitution - In")
        # INX / IOX = New scoreline
        if 'IB' in data or 'IF' in data:
            raw_min = data.get('IB', '0')
            minute_num = 0
            extra_min: Optional[int] = None
            
            # Parse minutes like "45+2'" or "78'"
            min_clean = raw_min.replace("'", "").strip()
            if '+' in min_clean:
                try:
                    m_parts = min_clean.split('+')
                    minute_num = int(m_parts[0])
                    extra_min = int(m_parts[1])
                except:
                    pass
            elif min_clean.isdigit():
                minute_num = int(min_clean)

            team_side = 'home' if data.get('IA') == '1' else 'away'
            player_name = data.get('IF', 'Player')
            detail_type = data.get('IK', '')
            
            # Normalize event type
            event_type = 'GOAL'
            detail = None
            lower_detail = detail_type.lower()
            
            if 'yellow card' in lower_detail:
                event_type = 'YELLOW_CARD'
            elif 'red card' in lower_detail or 'yellow/red' in lower_detail:
                event_type = 'RED_CARD'
                if 'yellow/red' in lower_detail:
                    event_type = 'YELLOW_RED_CARD'
            elif 'substitution' in lower_detail:
                event_type = 'SUBSTITUTION'
                detail = detail_type
            elif 'penalty' in lower_detail:
                event_type = 'GOAL'
                detail = 'Penalty'
            elif 'own goal' in lower_detail:
                event_type = 'GOAL'
                detail = 'Own Goal'
            elif 'assistance' in lower_detail:
                event_type = 'GOAL'
                detail = 'Assisted'
            else:
                event_type = 'GOAL'
                detail = detail_type or None

            # Scoreline
            home_sc: Optional[int] = None
            away_sc: Optional[int] = None
            if 'INX' in data and data['INX'].isdigit():
                home_sc = int(data['INX'])
            if 'IOX' in data and data['IOX'].isdigit():
                away_sc = int(data['IOX'])

            event_id = f"ev_{match_id}_{minute_num}_{team_side}_{re.sub(r'[^a-zA-Z0-9]', '_', player_name.lower())[:16]}"

            events.append({
                'id': event_id,
                'matchId': match_id,
                'type': event_type,
                'minute': minute_num,
                'extraMinute': extra_min,
                'teamSide': team_side,
                'playerName': player_name,
                'secondaryPlayerName': None,
                'detail': detail,
                'homeScore': home_sc,
                'awayScore': away_sc,
                'createdAt': datetime.datetime.now(datetime.timezone.utc).isoformat(),
            })

    return events

def parse_flashscore_statistics(match_id: str, raw_text: str) -> Dict[str, Any]:
    """
    Parse Flashscore's statistics feed (df_st_1_... format).
    """
    blocks = raw_text.split('¬~')
    stats: Dict[str, Any] = {}

    for block in blocks:
        if not block.strip():
            continue
        parts = block.split('¬')
        data = {}
        for p in parts:
            if '÷' in p:
                k, v = p.split('÷', 1)
                data[k] = v

        stat_name = data.get('SG', '').strip()
        val_home = data.get('SH', '').strip()
        val_away = data.get('SI', '').strip()

        if not stat_name:
            continue

        def parse_stat_num(s: str) -> Optional[int]:
            s_clean = s.replace('%', '').strip()
            if s_clean.isdigit():
                return int(s_clean)
            return None

        lower_name = stat_name.lower()
        if 'possession' in lower_name:
            stats['possessionHome'] = parse_stat_num(val_home)
            stats['possessionAway'] = parse_stat_num(val_away)
        elif 'total shots' in lower_name:
            stats['shotsHome'] = parse_stat_num(val_home)
            stats['shotsAway'] = parse_stat_num(val_away)
        elif 'shots on target' in lower_name:
            stats['shotsOnTargetHome'] = parse_stat_num(val_home)
            stats['shotsOnTargetAway'] = parse_stat_num(val_away)
        elif 'corner' in lower_name:
            stats['cornersHome'] = parse_stat_num(val_home)
            stats['cornersAway'] = parse_stat_num(val_away)
        elif 'fouls' in lower_name:
            stats['foulsHome'] = parse_stat_num(val_home)
            stats['foulsAway'] = parse_stat_num(val_away)
        elif 'yellow' in lower_name:
            stats['yellowCardsHome'] = parse_stat_num(val_home)
            stats['yellowCardsAway'] = parse_stat_num(val_away)
        elif 'red' in lower_name:
            stats['redCardsHome'] = parse_stat_num(val_home)
            stats['redCardsAway'] = parse_stat_num(val_away)
        elif 'offsides' in lower_name:
            stats['offsidesHome'] = parse_stat_num(val_home)
            stats['offsidesAway'] = parse_stat_num(val_away)
        elif 'saves' in lower_name:
            stats['savesHome'] = parse_stat_num(val_home)
            stats['savesAway'] = parse_stat_num(val_away)

    return stats

File: server/scraper/flashscore_parser.ts
import { Match, MatchEvent, MatchStats, MatchStatus, League, Team } from '../types.js';

export function parseFlashscoreFeed(rawText: string): Match[] {
  const blocks = rawText.split('¬~');
  const matches: Match[] = [];

  let currentLeague: League = {
    id: 'unknown',
    name: 'Football Match',
    country: 'World',
    countryCode: '',
  };

  for (const block of blocks) {
    if (!block.trim()) continue;

    const parts = block.split('¬');
    const data: Record<string, string> = {};
    for (const p of parts) {
      const idx = p.indexOf('÷');
      if (idx !== -1) {
        const k = p.slice(0, idx);
        const v = p.slice(idx + 1);
        data[k] = v;
      }
    }

    // Check if tournament header block
    if (data['ZA']) {
      const rawZa = data['ZA'];
      let country = data['ZY'] || 'World';
      let leagueName = rawZa;

      if (rawZa.includes(':')) {
        const partsZa = rawZa.split(':');
        country = partsZa[0].trim();
        leagueName = partsZa.slice(1).join(':').trim();
      }

      currentLeague = {
        id: data['ZC'] || data['ZEE'] || `league_${leagueName.toLowerCase().replace(/[^a-z0-9]/g, '_')}`,
        name: leagueName,
        country: country,
        countryCode: data['ZB'] || '',
        flag: data['OAJ'] || '',
      };
    } else if (data['AA']) {
      // Match block
      const matchId = data['AA'];
      const homeName = data['AE'] || 'Home Team';
      const awayName = data['AF'] || 'Away Team';

      const rawHomeScore = data['AG'] ?? '-';
      const rawAwayScore = data['AH'] ?? '-';
      const homeScore = /^\d+$/.test(rawHomeScore) ? parseInt(rawHomeScore, 10) : 0;
      const awayScore = /^\d+$/.test(rawAwayScore) ? parseInt(rawAwayScore, 10) : 0;

      const statusCode = data['AB'] || '1';
      const stageCode = data['AC'] || '';
      let statusText = stageCode;
      let status: MatchStatus = 'SCHEDULED';
      let minute: number | undefined = undefined;

      let startTime = new Date().toISOString();
      let startTimestampMs = 0;
      if (data['AD']) {
        try {
          const ts = parseInt(data['AD'], 10);
          if (!isNaN(ts)) {
            startTimestampMs = ts * 1000;
            startTime = new Date(startTimestampMs).toISOString();
          }
        } catch {
          // Keep current time
        }
      }

      let secondHalfStartMs = 0;
      if (data['AO']) {
        try {
          const ts = parseInt(data['AO'], 10);
          if (!isNaN(ts)) {
            secondHalfStartMs = ts * 1000;
          }
        } catch {}
      }

      if (statusCode === '2') {
        // Active / In Play / Halftime
        const scLower = stageCode.toLowerCase().trim();
        const isTrueHalftime =
          stageCode === '38' ||
          stageCode.toUpperCase() === 'HT' ||
          scLower === 'half time' ||
          scLower === 'halftime' ||
          scLower === 'half-time' ||
          scLower === 'break';

        if (isTrueHalftime && !scLower.includes('1st') && !scLower.includes('2nd')) {
          status = 'PAUSED';
          minute = 45;
          statusText = "HT (45')";
        } else if (stageCode === '6' || stageCode.includes('ET 1')) {
          status = 'EXTRA_TIME';
          minute = 95;
          statusText = "ET (1st Half)";
        } else if (stageCode === '7' || stageCode.includes('ET 2')) {
          status = 'EXTRA_TIME';
          minute = 110;
          statusText = "ET (2nd Half)";
        } else if (stageCode === '8' || stageCode.includes('Pen')) {
          status = 'PENALTIES';
          statusText = "Penalties";
        } else if (stageCode === '13' || scLower.includes('2nd')) {
          // 2nd Half (Flashscore stage code 13)
          status = 'IN_PLAY';
          if (secondHalfStartMs > 0) {
            const elapsed2nd = Math.floor((Date.now() - secondHalfStartMs) / 60000);
            minute = 45 + Math.max(1, elapsed2nd);
          } else if (startTimestampMs > 0) {
            const elapsedTotal = Math.floor((Date.now() - startTimestampMs) / 60000);
            minute = Math.max(46, elapsedTotal - 15);
          } else {
            minute = 46;
          }
          statusText = `${minute}' (2nd Half)`;
        } else if (stageCode === '12' || stageCode === '46' || stageCode === '1' || scLower.includes('1st')) {
          // 1st Half (Flashscore stage code 12 or 46)
          status = 'IN_PLAY';
          if (startTimestampMs > 0) {
            const elapsed = Math.floor((Date.now() - startTimestampMs) / 60000);
            minute = Math.max(1, elapsed);
          } else {
            minute = 1;
          }
          statusText = `${minute}' (1st Half)`;
        } else {
          status = 'IN_PLAY';
          const plusMatch = stageCode.match(/^(\d+)\+(\d+)'?$/);
          const minMatch = stageCode.match(/^(\d+)'?$/);
          if (plusMatch) {
            minute = parseInt(plusMatch[1], 10);
            statusText = `${plusMatch[1]}+${plusMatch[2]}'`;
          } else if (minMatch && parseInt(minMatch[1], 10) <= 120 && minMatch[1] !== '13' && minMatch[1] !== '12' && minMatch[1] !== '38') {
            minute = parseInt(minMatch[1], 10);
            statusText = `${minute}'`;
          } else if (startTimestampMs > 0) {
            const elapsed = Math.floor((Date.now() - startTimestampMs) / 60000);
            if (elapsed >= 1 && elapsed <= 130) {
              minute = elapsed;
              statusText = `${minute}'`;
            }
          } else {
            statusText = 'LIVE';
          }
        }
      } else if (statusCode === '3') {
        status = 'FINISHED';
        const plusMatch = stageCode.match(/(\d+)\+(\d+)/);
        if (plusMatch) {
          statusText = `${plusMatch[1]}+${plusMatch[2]}'`;
        } else {
          statusText = 'FT (90\')';
        }
        minute = 90;
      } else if (statusCode === '4' || statusCode === '9') {
        status = 'POSTPONED';
      } else if (statusCode === '5' || statusCode === '6') {
        status = 'CANCELLED';
      } else if (statusCode === '1') {
        status = 'SCHEDULED';
      }

      // Parse added / injury time
      let addedTime: number | undefined = undefined;
      if (data['DF'] && /^\d+$/.test(data['DF'])) {
        addedTime = parseInt(data['DF'], 10);
      } else {
        const pm = stageCode.match(/\+(\d+)/);
        if (pm) addedTime = parseInt(pm[1], 10);
      }

      // Parse 1st and 2nd half scores
      let periodScores: { half1Home?: number; half1Away?: number; half2Home?: number; half2Away?: number } | undefined = undefined;
      const h1H = data['BC'] !== undefined && /^\d+$/.test(data['BC']) ? parseInt(data['BC'], 10) : undefined;
      const h1A = data['BD'] !== undefined && /^\d+$/.test(data['BD']) ? parseInt(data['BD'], 10) : undefined;
      const h2H = data['BE'] !== undefined && /^\d+$/.test(data['BE']) ? parseInt(data['BE'], 10) : undefined;
      const h2A = data['BF'] !== undefined && /^\d+$/.test(data['BF']) ? parseInt(data['BF'], 10) : undefined;
      if (h1H !== undefined || h1A !== undefined || h2H !== undefined || h2A !== undefined) {
        periodScores = {
          half1Home: h1H,
          half1Away: h1A,
          half2Home: h2H,
          half2Away: h2A,
        };
      }

      const homeTeam: Team = {
        id: data['CX'] || `team_${homeName.toLowerCase().replace(/[^a-z0-9]/g, '_')}`,
        name: homeName,
      };

      const awayTeam: Team = {
        id: data['CY'] || `team_${awayName.toLowerCase().replace(/[^a-z0-9]/g, '_')}`,
        name: awayName,
      };

      let initialStats: MatchStats | undefined = undefined;
      const redCardsHome = data['DA'] ? parseInt(data['DA'], 10) : undefined;
      const redCardsAway = data['DB'] ? parseInt(data['DB'], 10) : undefined;
      if (redCardsHome !== undefined || redCardsAway !== undefined) {
        initialStats = {
          redCardsHome: redCardsHome || 0,
          redCardsAway: redCardsAway || 0,
        };
      }

      matches.push({
        id: matchId,
        provider: 'flashscore',
        league: { ...currentLeague },
        homeTeam,
        awayTeam,
        homeScore,
        awayScore,
        status,
        statusText: statusText || (status === 'IN_PLAY' ? (minute ? `${minute}'` : 'Live') : status === 'FINISHED' ? 'FT (90\')' : 'Upcoming'),
        minute,
        addedTime,
        periodScores,
        startTime,
        stats: initialStats,
        lastUpdated: new Date().toISOString(),
      });
    }
  }

  return matches;
}

export function parseFlashscoreEvents(matchId: string, rawText: string): MatchEvent[] {
  const blocks = rawText.split('¬~');
  const events: MatchEvent[] = [];

  for (const block of blocks) {
    if (!block.trim()) continue;

    const parts = block.split('¬');
    const data: Record<string, string> = {};
    for (const p of parts) {
      const idx = p.indexOf('÷');
      if (idx !== -1) {
        data[p.slice(0, idx)] = p.slice(idx + 1);
      }
    }

    if (data['IB'] || data['IF']) {
      const rawMin = data['IB'] || '0';
      let minuteNum = 0;
      let extraMin: number | undefined = undefined;

      const minClean = rawMin.replace("'", '').trim();
      if (minClean.includes('+')) {
        const mParts = minClean.split('+');
        minuteNum = parseInt(mParts[0], 10) || 0;
        extraMin = parseInt(mParts[1], 10) || undefined;
      } else if (/^\d+$/.test(minClean)) {
        minuteNum = parseInt(minClean, 10);
      }

      const teamSide: 'home' | 'away' = data['IA'] === '1' ? 'home' : 'away';
      const playerName = data['IF'] || 'Player';
      const detailType = data['IK'] || '';
      const lowerDetail = detailType.toLowerCase();

      let eventType: MatchEvent['type'] = 'GOAL';
      let detail: string | undefined = undefined;

      if (lowerDetail.includes('yellow card')) {
        eventType = 'YELLOW_CARD';
      } else if (lowerDetail.includes('red card') || lowerDetail.includes('yellow/red')) {
        eventType = lowerDetail.includes('yellow/red') ? 'YELLOW_RED_CARD' : 'RED_CARD';
      } else if (lowerDetail.includes('substitution')) {
        eventType = 'SUBSTITUTION';
        detail = detailType;
      } else if (lowerDetail.includes('penalty')) {
        eventType = 'GOAL';
        detail = 'Penalty';
      } else if (lowerDetail.includes('own goal')) {
        eventType = 'GOAL';
        detail = 'Own Goal';
      } else if (lowerDetail.includes('assistance')) {
        eventType = 'GOAL';
        detail = 'Assisted';
      } else {
        eventType = 'GOAL';
        detail = detailType || undefined;
      }

      let homeSc: number | undefined = undefined;
      let awaySc: number | undefined = undefined;
      if (data['INX'] && /^\d+$/.test(data['INX'])) homeSc = parseInt(data['INX'], 10);
      if (data['IOX'] && /^\d+$/.test(data['IOX'])) awaySc = parseInt(data['IOX'], 10);

      const eventId = `ev_${matchId}_${minuteNum}_${teamSide}_${playerName.toLowerCase().replace(/[^a-z0-9]/g, '_').slice(0, 16)}`;

      events.push({
        id: eventId,
        matchId,
        type: eventType,
        minute: minuteNum,
        extraMinute: extraMin,
        teamSide,
        playerName,
        detail,
        homeScore: homeSc,
        awayScore: awaySc,
        createdAt: new Date().toISOString(),
      });
    }
  }

  return events;
}

export function parseFlashscoreStatistics(rawText: string): MatchStats {
  const blocks = rawText.split('¬~');
  const stats: MatchStats = {};

  const hasMatchSection = rawText.includes('SE÷Match');
  let currentSection = hasMatchSection ? '' : 'Match';

  for (const block of blocks) {
    if (!block.trim()) continue;

    const seMatch = block.match(/SE÷([^¬~]+)/);
    if (seMatch) {
      currentSection = seMatch[1].trim();
    }

    // When a cumulative "Match" section is present in the feed, ignore subsequent
    // sub-sections (like "1st Half", "2nd Half") so full match totals are not overwritten
    if (hasMatchSection && currentSection !== 'Match') {
      continue;
    }

    const parts = block.split('¬');
    const data: Record<string, string> = {};
    for (const p of parts) {
      const idx = p.indexOf('÷');
      if (idx !== -1) {
        data[p.slice(0, idx)] = p.slice(idx + 1);
      }
    }

    const statName = (data['SG'] || '').trim();
    const valHome = (data['SH'] || '').trim();
    const valAway = (data['SI'] || '').trim();

    if (!statName) continue;

    const parseNum = (s: string): number | undefined => {
      const clean = s.replace('%', '').trim();
      return /^\d+$/.test(clean) ? parseInt(clean, 10) : undefined;
    };

    const lower = statName.toLowerCase();
    if (lower.includes('possession') && stats.possessionHome === undefined) {
      stats.possessionHome = parseNum(valHome);
      stats.possessionAway = parseNum(valAway);
    } else if ((lower.includes('total shots') || lower === 'shots') && stats.shotsHome === undefined) {
      stats.shotsHome = parseNum(valHome);
      stats.shotsAway = parseNum(valAway);
    } else if ((lower.includes('shots on target') || lower.includes('shots on goal')) && stats.shotsOnTargetHome === undefined) {
      stats.shotsOnTargetHome = parseNum(valHome);
      stats.shotsOnTargetAway = parseNum(valAway);
    } else if (lower.includes('corner') && stats.cornersHome === undefined) {
      stats.cornersHome = parseNum(valHome);
      stats.cornersAway = parseNum(valAway);
    } else if (lower.includes('foul') && stats.foulsHome === undefined) {
      stats.foulsHome = parseNum(valHome);
      stats.foulsAway = parseNum(valAway);
    } else if (lower.includes('yellow') && stats.yellowCardsHome === undefined) {
      stats.yellowCardsHome = parseNum(valHome);
      stats.yellowCardsAway = parseNum(valAway);
    } else if (lower.includes('red') && stats.redCardsHome === undefined) {
      stats.redCardsHome = parseNum(valHome);
      stats.redCardsAway = parseNum(valAway);
    } else if (lower.includes('offside') && stats.offsidesHome === undefined) {
      stats.offsidesHome = parseNum(valHome);
      stats.offsidesAway = parseNum(valAway);
    } else if (lower.includes('save') && stats.savesHome === undefined) {
      stats.savesHome = parseNum(valHome);
      stats.savesAway = parseNum(valAway);
    } else if (lower.includes('substitution') && stats.substitutionsHome === undefined) {
      stats.substitutionsHome = parseNum(valHome);
      stats.substitutionsAway = parseNum(valAway);
    } else if (lower.includes('penalt') && stats.penaltiesHome === undefined) {
      stats.penaltiesHome = parseNum(valHome);
      stats.penaltiesAway = parseNum(valAway);
    }
  }

  return stats;
}
